### PR TITLE
Change Image parsing to Imaging instead of MediaInfo and Sanselan.

### PR DESCRIPTION
**Pull request by valib**
MediaInfo doesn't provide sufficient information and Sanselan is quite old so the follower Imaging is used.

---
This change is [Reviewable](https://reviewable.io/reviews/universalmediaserver/universalmediaserver/911)

### CODE REVIEW

**Comment by valib** [comment]
@UniversalMediaServer/developers and especially @Sami32 can you test it? it should solve also the issue  #910

**Comment by Sami32** [comment]
@valib Cool :-)
I will enjoy to test it as soon as the Snapshot will be disponible.
Thank you for time and work :+1: 

If i can make an suggestion, for debug purpose or later upgrade, maybe add `isProgressive` data in the _DLNAMediaInfo.java_ file ?

**Comment by valib** [comment]
@Sami32 you can try this http://www.universalmediaserver.com/forum/viewtopic.php?f=25&t=1782

**Comment by Sami32** [comment]
@valib Thank for your help, but when i tried i get so many errors and warning about Eclipse, Maven...so much verbose that i was not able to read as fast and understand all :-(

**Comment by Sami32** [comment]
@valib Not sure it can help you, but when i tried to compile with Netbeans i got this error concerning DLNAresource; so i'm not sure it's come from your commits or not...

> Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.19:test (default-test) on project ums: Execution default-test of goal org.apache.maven.plugins:maven-surefire-plugin:2.19:test failed: There was an error in the forked process
> java.lang.ClassFormatError: Absent Code attribute in method that is not native or abstract in class file net/pms/dlna/DLNAResource
> at java.lang.ClassLoader.defineClass1(Native Method)
> at java.lang.ClassLoader.defineClass(ClassLoader.java:800)
> at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
> at java.net.URLClassLoader.defineClass(URLClassLoader.java:449)
> at java.net.URLClassLoader.access$100(URLClassLoader.java:71)
> at java.net.URLClassLoader$1.run(URLClassLoader.java:361)
> at java.net.URLClassLoader$1.run(URLClassLoader.java:355)
> at java.security.AccessController.doPrivileged(Native Method)
> at java.net.URLClassLoader.findClass(URLClassLoader.java:354)
> at java.lang.ClassLoader.loadClass(ClassLoader.java:425)
> at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:308)
> at java.lang.ClassLoader.loadClass(ClassLoader.java:358)
> at java.lang.Class.getDeclaredMethods0(Native Method)
> at java.lang.Class.privateGetDeclaredMethods(Class.java:2625)
> at java.lang.Class.getMethod0(Class.java:2866)
> at java.lang.Class.getMethod(Class.java:1676)
> at org.apache.maven.surefire.util.ReflectionUtils.tryGetMethod(ReflectionUtils.java:61)
> at org.apache.maven.surefire.common.junit3.JUnit3TestChecker.isSuiteOnly(JUnit3TestChecker.java:64)
> at org.apache.maven.surefire.common.junit3.JUnit3TestChecker.isValidJUnit3Test(JUnit3TestChecker.java:59)
> at org.apache.maven.surefire.common.junit3.JUnit3TestChecker.accept(JUnit3TestChecker.java:54)
> at org.apache.maven.surefire.common.junit4.JUnit4TestChecker.accept(JUnit4TestChecker.java:52)
> at org.apache.maven.surefire.util.DefaultScanResult.applyFilter(DefaultScanResult.java:98)
> at org.apache.maven.surefire.junit4.JUnit4Provider.scanClassPath(JUnit4Provider.java:294)
> at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:138)
> at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:286)
> at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:240)
> at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:121)
> -> [Help 1]
> 
> To see the full stack trace of the errors, re-run Maven with the -e switch.
> Re-run Maven using the -X switch to enable full debug logging.
> 
> For more information about the errors and possible solutions, please read the following articles:

**Comment by ExSport** [comment]
Why you compile it manually?
Read the post from Valib about UMSBuilder.
It is easy like downloading, unpacking, starting bat file and that's all. Nothing more needed.

**Comment by Sami32** [comment]
@ExSport Because i get this error https://github.com/UniversalMediaServer/UniversalMediaServer/pull/911#issuecomment-224311142
So, i don't know how to get it work. Never mind, more skilled people will test it.

**Comment by ExSport** [comment]
@Sami32 Just wanted to write it your internet is blocked or proxy used:)
If it will not work for you I just compiled it so I can upload it somewhere but I am sure you already did it :+1:

**Comment by ExSport** [comment]
@valib By accident I compled another commit so when I tried this one, it can't be compiled due to the missing dependency:
```
[ERROR] Failed to execute goal on project ums: Could not resolve dependencies for project net.pms:ums:jar:6.3.3-SNAPSHOT: Could not find artifact org.apache.commons:commons-imaging:jar:1.0-SNAPSHOT
```
I didn't find it here in version 1.0 for download: https://commons.apache.org/proper/commons-imaging/

**Comment by valib** [comment]
@ExSport thanks for the comment.

**Comment by Sami32** [comment]
@ExSport I accept your offer, thank you :-)
So if you have a JAR link of your compilation, i will compare to mine, because when i tried my compilation, without using _Cache_ i didn't saw any differences...but i am under skilled so...

And you, do you get thumbnails for _TIF/TIFF_ images ? TIF are parsed as `image/jpeg` ?

**Comment by Sami32** [comment]
@valib It solved the TIF/TIFF error `java.lang.NullPointerException: Image cannot be null`. :+1:

**Comment by Sami32** [comment]
@valib I saw that you add **BMP** in _ImagesUtil.java_, and since we can find it in some renderer, do you think it will be correct to add it as well to your commit https://github.com/UniversalMediaServer/UniversalMediaServer/pull/911/commits/c18e7a216b589b5996df7f1071bdfc6ccd447312 ?
```
} else if ("bmp".equals(codecV) || "bmp".equals(container)) {
+					mimeType = HTTPResource.BMP_TYPEMIME;
```

**Comment by Sami32** [comment]
@valib 

> TRACE 2016-06-15 03:54:02.203 [Library Scanner] Matched format TIF to "C:\Program Files\Universal Media Server\win32\FFMPEG_Fix\123.tif"
> TRACE 2016-06-15 03:54:02.484 [Library Scanner] Creating (temporary) thumbnail: 123.tif
> DEBUG 2016-06-15 03:54:02.546 [Library Scanner] Failed to resize image: Image cannot be null.
> TRACE 2016-06-15 03:54:02.546 [Library Scanner] 
> java.lang.NullPointerException: Image cannot be null.
> 	at net.coobird.thumbnailator.tasks.io.BufferedImageSource.<init>(Unknown Source)
> 	at net.coobird.thumbnailator.Thumbnails$Builder$BufferedImageImageSourceIterator$1.next(Unknown Source)
> 	at net.coobird.thumbnailator.Thumbnails$Builder$BufferedImageImageSourceIterator$1.next(Unknown Source)
> 	at net.coobird.thumbnailator.Thumbnails$Builder.toOutputStream(Unknown Source)
> 	at net.pms.util.UMSUtils.scaleImage(UMSUtils.java:637)
> 	at net.pms.dlna.DLNAMediaInfo.parse(DLNAMediaInfo.java:885)
> 	at net.pms.dlna.DLNAMediaInfo.generateThumbnail(DLNAMediaInfo.java:524)
> 	at net.pms.dlna.DLNAResource.checkThumbnail(DLNAResource.java:3177)
> 	at net.pms.dlna.RealFile.checkThumbnail(RealFile.java:289)
> 	at net.pms.dlna.RealFile.isValid(RealFile.java:86)
> 	at net.pms.dlna.DLNAResource.addChild(DLNAResource.java:561)
> 	at net.pms.dlna.DLNAResource.addChild(DLNAResource.java:521)
> 	at net.pms.dlna.MapFile.manageFile(MapFile.java:114)
> 	at net.pms.dlna.MapFile.analyzeChildren(MapFile.java:171)
> 	at net.pms.dlna.RootFolder.scan(RootFolder.java:259)
> 	at net.pms.dlna.RootFolder.scan(RootFolder.java:216)
> 	at net.pms.dlna.DLNAMediaDatabase.run(DLNAMediaDatabase.java:814)
> 	at java.lang.Thread.run(Unknown Source)

**Comment by valib** [comment]
@UniversalMediaServer/developers is it working for you?

**Comment by SubJunk** [comment]
@valib I'll test it this week

**Comment by SubJunk** [comment]
@valib tests are failing in FileUtilTest

**Comment by SubJunk** [comment]
@valib I just tested it and if I browse images on my TV it has thumbnails for PNG images but I can't click them, as before. How do you view unsupported images?

**Comment by Sami32** [comment]
Testing TIFF samples :
[libtiffpic.7z.txt](https://github.com/UniversalMediaServer/UniversalMediaServer/files/337752/libtiffpic.7z.txt)

**Comment by valib** [comment]
I divided that PR and reverted the part with images transcoding to be able to merge it to master. The transcoding part is now in https://github.com/UniversalMediaServer/UniversalMediaServer/tree/image_transcode

**Comment by Sami32** [comment]
@valib It's look very good already :-)

File: src/main/external-resources/renderers/Panasonic-VieraTXL32V10E.conf [hunk 1]
```diff
@@ -41,9 +41,11 @@ Supported = f:mpegps     v:mpeg2|mp4|h264   a:ac3|lpcm|aac|mpa   m:video/mpeg
...
 
 # Supported image formats:
-Supported = f:jpg   m:image/jpeg
-Supported = f:png   m:image/png
-Supported = f:gif   m:image/gif
+Supported = f:jpg     	m:image/jpeg
+Supported = f:png		m:image/png
+Supported = f:gif		m:image/gif
+Supported = f:tiff		m:image/tiff
+Supported = f:bmp		m:image/bmp
```
Comment (valib) [inline src/main/external-resources/renderers/Panasonic-VieraTXL32V10E.conf]: @SubJunk when my old Panasonic TV supports that not documented images formats  what about to add this setting to the other Panasonic TVs confs? Users can test it and we can remove the not working formats for particular renderers.

**Comment by atamariya** [comment]
I'm stumbled across this issue while working for my zeroconf changes. Looks like there are just too many libraries in use or lying around.

`DLNAMediaInfo.getFFmpegThumbnail(inputFile, resume, renderer);`

@valib This can pretty much handle multiple formats. May be it's better to fix the code to use this rather than introducing another library.

**Comment by valib** [comment]
@atamariya sorry for that late comment but I forgot to comment emmediately. I tested the FFmpeg support and I must say that to implement it as you requested is a problem. It could be possible to use FFmpeg to parse medias but in that case we should resign on some implemented features. E.g. FFmpeg doesn't support extracting CoverArt for audio files so the jaudiotagger is used to do it.  And when it is used it is logical to use it to parse all other information from the file tag. The same applies for images when FFmpeg and also MediaInfo.dll doesn't provide the EXIF information like the ISO, camera model or exposure time and the Sanselan or the follower Imaging is used to do it.

**Comment by Sami32** [comment]
Reviewed 2 of 5 files at r2, 1 of 16 files at r10, 7 of 16 files at r11, 1 of 1 files at r12.
Review status: 5 of 9 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/universalmediaserver/universalmediaserver/911)*

**Comment by Sami32** [comment]
*[src/main/external-resources/renderers/Panasonic-VieraTXL32V10E.conf, line 49 at r12](https://reviewable.io:443/reviews/universalmediaserver/universalmediaserver/911#-KKAnVIX9bDcgapdbikU-r12-49:-KUPZX6Y_nZkDmVdXSJb:b-qce0ai) ([raw file](https://github.com/universalmediaserver/universalmediaserver/blob/50398f860a17ce3166e9ed0d8a9d761ef86efce2/src/main/external-resources/renderers/Panasonic-VieraTXL32V10E.conf#L49)):*
<details><summary><i>Previously, valib (Libor Vala) wrote…</i></summary>
> @SubJunk when my old Panasonic TV supports that not documented images formats  what about to add this setting to the other Panasonic TVs confs? Users can test it and we can remove the not working formats for particular renderers.
</details>
Not mine, i tested long ago

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/universalmediaserver/universalmediaserver/911)*

**Comment by Sami32** [comment]
LGBM

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/universalmediaserver/universalmediaserver/911#-:-KUPa86sRM9SXfv5T0Se:bb6yw7f)*

**Comment by Sami32** [comment]
LGTM

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/universalmediaserver/universalmediaserver/911#-:-KUPabJNUflnoSXEABSX:bb6ywmx)*

**Comment by Sami32** [comment]
:lgtm:

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/universalmediaserver/universalmediaserver/911#-:-KUPageITLmKhOyyfrSC:bnfp4nl)*

**Comment by Sami32** [comment]
Reviewed 1 of 6 files at r13, 3 of 4 files at r14.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/universalmediaserver/universalmediaserver/911)*

**Comment by Sami32** [comment]
@valib I leave it to you to delete this branch as i don't make to much mess...

**Comment by Nadahar** [comment]
@Sami32 Did you mean to merge it?

**Comment by Sami32** [comment]
@valib If you asked me if i choice to merge it, it is yes.
As i understood that @valib was waiting for that and that nobody cared.
I hope only i didn't giving again a mess to him 8-/

**Comment by Sami32** [comment]
@valib Do i need to revert this merge ?

**Comment by valib** [comment]
@Sami32 it is OK. If could be reverted later if somebody will repport any problem.

**Comment by Nadahar** [comment]
@Sami32 Generally all this (including the requirement for one approved review) is about not merging things that might need to be reversed. The reason is that it creates a lot of mess in the history, and a lot of unnecessary sync/merges into other PRs. It's just much easier to try to make sure that what is merged is final and doesn't need any more fixes or adjustments.

Another thing is if the change is something that is agreed upon. Some changes might not seem like such a good idea for others as for the author.

You said yourself above that you don't fully understand the PR and have just tested it. Whenever that is the case you should NOT merge something imo. It's good that you tested it, you should then just submit a review saying that you tested it and didn't find anything, and then let someone that fully understand the PR evaluate if it should actually be merged, given that the change itself is agreed upon.

**Comment by Sami32** [comment]
@Nadahar Waiting evaluation for how long ?
Since i read your complains about no one care or evaluate your PR, i'm just curious what should be the "correct" evaluation time waiting ?

**Comment by Nadahar** [comment]
@Sami32 The problem is that this isn't up to me, I can't dictate the rules. I think there should be a system for this, and I agree that it can be very frustrating not getting any feedback. At the same time just pushing changes into master unreviewed isn't a solution - it's worse imo.

Lacking a proper procedure for this, I've sometimes said that unless someone objects within a day or so I'm going to merge - and then merged. While this is far from perfect, at least it gives others a chance to stop it if they feel they need to.

**Comment by Sami32** [comment]
@Nadahar I find 1 day or so is really too short imo, 1 week should be more safe and respectuous of the planning or daytime difference of the others developpers, but 20 days is long enough...so if i agree and get no resting errors i will approve the PRs without merge it if i don't understood all.

**Comment by valib** [comment]
@UniversalMediaServer/developers I agree that one week or better two weeks should be the trigger to merge  PR's when no objestions are raised. I believe that all of us are testing their PR's beforre they are commited.

**Comment by Nadahar** [comment]
As I see it, there's several things that should be approved but not everything apply to every PR:

1. The changes to features/functionality (if it applies)
2. The quality/compatibility of the code
3. Testing results (if it is relevant)

Is I see it some of the problem that it's hard to indicate what you actually approve. For example I might look at what a PR does and tests it without problems, and want to indicate that I approve 1) and 3) but I haven't taken the time to study 2). In another case I might have studied the code, but isn't sure that the change is good - so I only want to indicate that I approve 2). Currently I think it's difficult to differentiate between these things when I review something, because it isn't properly defined what a review is approving.

Not every PR has all these, but everything has 2). I think 1) and 3) should also be "approved" when they apply. I don't think everybody has to approve every part, one might approve 1) and 3) while another approve 2). I think there should be a consensus about 1), while it will often be sufficient that one other person has reviewed 2) and 3) - depending on how big and "risky" the changes are.

I know that you guys don't want to much bureaucracy and I don't have any practical solution for how to make this work effectively, but those are the things I think should be considered before a PR is merged.

When I create a PR I often start with writing a descriptive text about what my goal is, and I might continue to push commits for a while after that. My motivation for doing this is to get some feedback for 1), because if there is no consensus about 1) there's no need for me to do the work. If 1) is "approved" and I finish and test the PR, I generally make a note that "I now think this is ready for merge". That actually means that now I want feedback on 2) and 3).

Small "bugfix" and other fixes are easy, as only 2) need to be reviewed.

@DevlshOne Do any of you have any thoughts about this?

**Comment by Nadahar** [comment]
@UniversalMediaServer/developers This PR has turned out to create problems while building at Travis. Many times the build stalls or fails because the Apache Snapshot Repository is unavailable / doesn't reply, and this leads to very long build times and/or failed Travis builds.

Generally I want to say that we should be extremely cautios with using ```SNAPSHOT``` versions as dependencies. They are not meant to be used for other than testing. The reason for this is the way Maven versioning works: Any version not containing ```SNAPSHOT``` is considered a release and is final/can't be updated. Any version containing ```SNAPSHOT``` is the same as a "nightly build", it is in whatever state the code was in the last time it was pushed to Maven. They can change without notice, method calls can change, functionality can change and new bugs can appear - as in all code being developed. When using snapshots we have no control over this, neither do we get any warning or information that indicate that anything has changed. It's basicly the same as for our master snapshots inbetween releases. Things can break at any second. 

Maven central repository, which has several replicates and is generally supposed to be available at any time, does not allow ```SNAPSHOT``` versions to be published exactly because they are unstable. That means that in addition to the uncertainty tied to using "live" code we have to rely on other repositories for snapshots. Those don't have redundancy/replicas and has no guarantee of being available.

In short, it is my view that ```SNAPSHOT``` versions should be avoided at almost any cost. Regarding Imaging I figured I'll have to replace the ```SNAPSHOT``` version with the latest stable, and went to look for it. This is what I found:
```
Help Needed!
There is currently no stable release of Imaging available, because there are a lot of outstanding issues. If you want to help getting out a release you may:

    Look through the issues waiting for a patch
    Look through the Needs Work tags in the TagList report
    Contribute a patch
    Create a PR via Github

Make sure to read through our guide lines for contributing patches before you start coding. This will stream line the process of getting your contributions into the code.
Latest Imaging SNAPSHOT

There is currently no stable release of Imaging. You may:

    Build Imaging from sources by checking out the source code from the SVN repository.
    Pull the latest snapshot from the Apache snapshot repository
```

So basicly, there is no release of ```commons-imaging``` because it is not stable and has a lot of issues. It doesn't sound like something we should use at all to be honest.

I don't know what we should do about this, but it can't stay like it is with builds frequenty failing. I'm also skeptical of using a library that is deemed to be unstable and unfinished by its authors.

**Comment by Sami32** [comment]
@Nadahar We all know your opinion on snapshots, but:
`Sanselan` 0.97 don't have TIFF or progressive JPEG support for example...
`commons-imaging` receive regular update. So it's not perfect or ideal, but as we do already with the other snapshot dependancies, seem to me a reasonnable risk/advantage compromise. It's only my opinion though...

For a more reliable repository access why not just change the download link with something like these ones ?
http://repository.jboss.org/nexus/content/groups/public/org/apache/commons/commons-imaging/1.0-SNAPSHOT/
https://repository.mulesoft.org/nexus/content/repositories/public/org/apache/commons/commons-imaging/1.0-SNAPSHOT/
http://maven.wso2.org/nexus/content/groups/wso2-public/org/apache/commons/commons-imaging/1.0-SNAPSHOT/

**Comment by valib** [comment]
Agree that using snapshots is dangerous but it was the only way to implement support for images @Sami32 mentioned. Currently the Imaging snapshot is not available in the repostory so I raised the issue in the Imaging issue tracker.

**Comment by Nadahar** [comment]
@Sami32 If this is just "my optionion" I'm not sure you understand the consequence. My main point isn't the fact that it isn't "release quality", but that it will change without notice meaning that our code can break from one day to the other "by itself". If this is the only thing available it would be better for us to make a copy of the artifact and provide it some other way (e.g via the external plugin). That way it would at least be stable in that it wouldn't change until we did something.

The problem with adding Maven repositories is how Maven work. For some reason it insists on querying every repository for every artifact, when means that it's not only imaging this impact but all our Maven dependencies. I've watched it during builds and this repository generally takes a long time to reply for requests for other artifacts as well, and sometimes it fails to reply and the build fails with a timeout. Another factor that only applies to Travis builds is that Travis has their own local Maven central replica - which means that anything from central is lightning quick while all other repos have to be contacted a lot of times for each build.

**Comment by Nadahar** [comment]
@valib What about [Apache Tika](https://tika.apache.org/)? It can read [metadata from TIFFs and JPGs](https://tika.apache.org/1.13/api/org/apache/tika/parser/image/package-summary.html). 

Tika is [available](http://mvnrepository.com/artifact/org.apache.tika/tika/1.13) at Maven central and has its own [artifact for just the parsers](http://mvnrepository.com/artifact/org.apache.tika/tika-parsers/1.13). I have never tested it though, so I have no idea if it can do all we need.

**Comment by Sami32** [comment]
@Nadahar "optionion" ??? it's already difficult for me to understand so if you make your own word...
I don't have any problem if you feel more safe with having our own local copy.

I saw this timeout issue but supposed that a different download repository link will solve it. Btw the way it was fine some weeks ago.

EDIT: Tika is a big, i can't remember how many MB though.

**Comment by Nadahar** [comment]
I honestly don't know exactly how we're using it. Is it used only for parsing/detection or is it also used to "transcode" images? Tika is parsing only as far as I can see, while Imaging can convert images etc.

**Comment by Nadahar** [comment]
@UniversalMediaServer/developers Are we just going to leave it like this with extremely slow builds that fail randomly? I suspect that posting an issue won't change anything, snapshot repositories aren't supposed to be generally available and artifacts can disappear and reappear without notice.

**Comment by SubJunk** [comment]
Maybe it should be reverted

**Comment by Nadahar** [comment]
@SubJunk At the very least another repository that is stable should be found. You could simply host the jar and add it as an "external" dependency. That way it would be installed during external:install and the artifact wouldn't suddenly change or become unavailable.

**Comment by valib** [comment]
@Nadahar @SubJunk I think that this is a good idea to host the jar and add it as an "external" dependency. I hope that in the near future they will release the stable version and we will stop using this SNAPSHOT.

**Comment by Nadahar** [comment]
@valib Yeah, that's probably a good idea given that all the work is already done and everything is merged. It's very easy to set up the download in the POM, but @SubJunk need to provide an URL for the JAR.

**Pull request merged**